doc/dplus.1
.TH dplus 1 "May 11, 2012" "version 0.5" "USER COMMANDS"
.SH NAME
dplus \- web browser
.SH SYNOPSIS
.B dplus
.RI [ OPTION ]...
.RB [ \-\- ]
.RI [ URL | FILE ]...
.SH DESCRIPTION
.PP
DPlus is a graphical web browser with an emphasis on security, performance,
and portability.
.SH OPTIONS
.TP
\fB\-f\fR, \fB\-\-fullwindow\fR
Start in full window mode: hide address bar, navigation buttons, menu, and
status bar.
.TP
\fB\-g\fR, \fB\-\-geometry \fIGEO\fR
Set initial window position where \fIGEO\fR is
\fIW\fBx\fIH\fR[{\fB+\-\fR}\fIX\fR{\fB+\-\fR}\fIY\fR].
.TP
\fB\-h\fR, \fB\-\-help\fR
Display this help text and exit.
.TP
\fB\-l\fR, \fB\-\-local\fR
Don't load images for these URL(s).
.TP
\fB\-v\fR, \fB\-\-version\fR
Display version info and exit.
.TP
\fB\-x\fR, \fB\-\-xid \fIXID\fR
Open first DPlus window in an existing window whose window ID is \fIXID\fR.
.SH EXIT STATUS
.TP
.B 0
No error.
.TP
.B 1
Internal error.
.TP
.B 2
Error in command line arguments.
.SH ENVIRONMENT
.TP
.BR "HOME " "(or " "HOMEDRIVE " "and " "HOMEPATH " "on Cygwin)"
User's home directory.
.TP
.B http_proxy
URL of proxy to send HTTP traffic through.
.SH FILES
.TP
.I ~/.dplus/bookmark.txt
User bookmarks
.TP
.I ~/.dplus/certs/
Saved certificates for HTTPS.
.TP
.I ~/.dplus/cookiesrc
Cookie settings
.TP
.I ~/.dplus/dplusrc
Configuration file.
.TP
.I ~/.dplus/keysrc
Keybindings.
.TP
.I ~/.dplus/style.css
User style sheet
.SH SEE ALSO
.BR D+ Browser homepage:
.B http://dplus-browser.sourceforge.net/


ChangeLog
----------------------
 D+ Browser ChangeLog
----------------------

This ChangeLog provides a broad overview of changes between releases.
Please see the Mercurial commit log for more specific details.


Version 0.5 (not released yet)
------------------------------

General:
- Initial release (forked from Dillo-3.0).
- Add support for Microsoft Windows and MS-DOS systems.
- Remove legacy DPI (Dillo plugins) code.

Network and Input/Output:
- Build cookies support into the browser.
- Build HTTPS support into the browser.
- Build file protocol support into the browser.
- Minor bugfixes.

Preferences:
- Add support for running as a portable application on Microsoft Windows.
- Follow the local platform's naming conventions for configuration files.
- Allow overriding the bookmarks file location via prefs.bookmarks_file.

User Interface:
- Build bookmarks support into the browser.
- Add a graphical interface for setting program preferences.
- Add preliminary printing support.
- Use native file dialogs when they're available.
- Use system colors rather than a hard-coded color scheme.
- Improve clipboard support.
- Don't use Unix key bindings in text input boxes.
- Improve the toolbar's appearance and usability.
- Replace the search dialog with a Firefox-style search bar.
- Implement find-as-you-type and provide visual feedback in the find bar.
- Treat Ctrl-clicking a hyperlink the same as middle-clicking it.
- Use standard keyboard shortcuts to match other browsers.
- Always show the tab bar by default.
- Add a zoom control in the status bar.
- Add an "Add search engine..." option to the form right-click menu.
- Minor bugfixes.


src/dplus.cc
/*
 * ------------
 *  D+ Browser
 * ------------
 *
 * Copyright 1999-2007 Jorge Arellano Cid <jcid@dillo.org>
 * Copyright 2012 Benjamin Johnson <obeythepenguin@users.sourceforge.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <signal.h>
#include <locale.h>

#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/fl_draw.H>

#include "msg.h"
#include "paths.hh"
#include "uicmd.hh"

#include "prefs.h"
#include "prefsparser.hh"
#include "keys.hh"
#include "bw.h"
#include "misc.h"
#include "history.h"

#include "dns.h"
#include "web.hh"
#include "IO/Url.h"
#include "IO/mime.h"
#include "capi.h"
#include "dicache.h"
#include "cookies.h"
#include "bookmark.hh"
#include "auth.h"
#include "download.hh"
#include "unicows.h"
#include "file.h"

#include "dw/fltkcore.hh"

#include "../dlib/dsock.h"

/*
 * Command line options structure
 */
typedef enum {
   DILLO_CLI_NONE          = 0,
   DILLO_CLI_XID           = 1 << 0,
   DILLO_CLI_FULLWINDOW    = 1 << 1,
   DILLO_CLI_HELP          = 1 << 2,
   DILLO_CLI_VERSION       = 1 << 3,
   DILLO_CLI_LOCAL         = 1 << 4,
   DILLO_CLI_GEOMETRY      = 1 << 5,
   DILLO_CLI_ERROR         = 1 << 15,
} OptID;

typedef struct {
   const char *shortopt;
   const char *longopt;
   int opt_argc; /* positive: mandatory, negative: optional */
   OptID id;
   const char *help;
} CLI_options;

static const CLI_options Options[] = {
   {"-f", "--fullwindow", 0, DILLO_CLI_FULLWINDOW,
    "  -f, --fullwindow       Start in full window mode: hide address bar,\n"
    "                         navigation buttons, menu, and status bar."},
   {"-g", "-geometry",    1, DILLO_CLI_GEOMETRY,
    "  -g, -geometry GEO      Set initial window position where GEO is\n"
    "                         WxH[{+-}X{+-}Y]"},
   {"-h", "--help",       0, DILLO_CLI_HELP,
    "  -h, --help             Display this help text and exit."},
   {"-l", "--local",      0, DILLO_CLI_LOCAL,
    "  -l, --local            Don't load images for these URL(s)."},
   {"-v", "--version",    0, DILLO_CLI_VERSION,
    "  -v, --version          Display version info and exit."},
   {"-x", "--xid",        1, DILLO_CLI_XID,
    "  -x, --xid XID          Open first Dillo window in an existing\n"
    "                         window whose window ID is XID."},
   {NULL, NULL, 0, DILLO_CLI_NONE, NULL}
};

/*
 * Print help text generated from the options structure
 */
static void printHelp(const char *cmdname, const CLI_options *options)
{
   printf("Usage: %s [OPTION]... [--] [URL|FILE]...\n"
          "Options:\n", cmdname);
   while (options && options->help) {
      printf("%s\n", options->help);
      options++;
   }
   printf("  URL                    URL to browse.\n"
          "  FILE                   Local FILE to view.\n"
          "\n");
}

/*
 * Return the maximum number of option arguments
 */
static int numOptions(const CLI_options *options)
{
   int i, max;

   for (i = 0, max = 0; options[i].shortopt; i++)
      if (abs(options[i].opt_argc) > max)
         max = abs(options[i].opt_argc);
   return max;
}

/*
 * Get next command line option.
 */
static OptID getCmdOption(const CLI_options *options, int argc, char **argv,
                           char **opt_argv, int *idx)
{
   typedef enum { O_SEARCH, O_FOUND, O_NOTFOUND, O_DONE } State;
   OptID opt_id = DILLO_CLI_NONE;
   int i = 0;
   State state = O_SEARCH;

   if (*idx >= argc) {
      state = O_DONE;
   } else {
      state = O_NOTFOUND;
      for (i = 0; options[i].shortopt; i++) {
         if (strcmp(options[i].shortopt, argv[*idx]) == 0 ||
             strcmp(options[i].longopt, argv[*idx]) == 0) {
            state = O_FOUND;
            ++*idx;
            break;
         }
      }
   }
   if (state == O_FOUND) {
      int n_arg = options[i].opt_argc;
      opt_id  = options[i].id;
      /* Find the required/optional arguments of the option */
      for (i = 0; *idx < argc && i < abs(n_arg) && argv[*idx][0] != '-'; i++)
         opt_argv[i] = argv[(*idx)++];
      opt_argv[i] = NULL;

      /* Optional arguments have opt_argc < 0 */
      if (i < n_arg) {
         fprintf(stderr, "Option %s requires %d argument%s\n",
                 argv[*idx-i-1], n_arg, (n_arg == 1) ? "" : "s");
         opt_id = DILLO_CLI_ERROR;
      }
   }
   if (state == O_NOTFOUND) {
      if (strcmp(argv[*idx], "--") == 0)
         (*idx)++;
      else if (argv[*idx][0] == '-') {
         fprintf(stderr, "Command line option \"%s\" not recognized.\n",
                 argv[*idx]);
         opt_id = DILLO_CLI_ERROR;
      }
   }
   return opt_id;
}

/*
 * Set FL_FREE_LABELTYPE to interpret neither symbols (@) nor shortcuts (&).
 */
static void custLabelDraw(const Fl_Label* o, int X, int Y, int W, int H,
                          Fl_Align align)
{
   const int interpret_symbols = 0;

   fl_draw_shortcut = 0;
   fl_font(o->font, o->size);
   fl_color((Fl_Color)o->color);
   fl_draw(o->value, X, Y, W, H, align, o->image, interpret_symbols);
}

static void custLabelMeasure(const Fl_Label* o, int& W, int& H)
{
   const int interpret_symbols = 0;

   fl_draw_shortcut = 0;
   fl_font(o->font, o->size);
   fl_measure(o->value, W, H, interpret_symbols);
}

/*
 * Replace FL_DIAMOND_BOX with a special box type for CustTabButton.
 */
static void custTabButtonUpBoxDraw(int x, int y, int w, int h, Fl_Color c)
{
   fl_color(c);
   fl_rectf(x, y, w, h);
   w--, h--;

   fl_color(FL_FOREGROUND_COLOR);
   fl_line(x+w, y, x+w, y+h);
}

static void custTabButtonDownBoxDraw(int x, int y, int w, int h, Fl_Color c)
{
   custTabButtonUpBoxDraw(x, y, w, h, fl_lighter(c));
}

/*
 * Tell the user if default/pref fonts can't be found.
 */
static void checkFont(const char *name, const char *type)
{
   if (! dw::fltk::FltkFont::fontExists(name))
      MSG_WARN("preferred %s font \"%s\" not found.\n", type, name);
}

static void checkPreferredFonts()
{
   checkFont(prefs.font_sans_serif, "sans-serif");
   checkFont(prefs.font_serif, "serif");
   checkFont(prefs.font_monospace, "monospace");
   checkFont(prefs.font_cursive, "cursive");
   checkFont(prefs.font_fantasy, "fantasy");
}

/*
 * Given a command line argument, build a DilloUrl for it.
 */
static DilloUrl *makeStartUrl(char *str, bool local)
{
   char *url_str, *p;
   DilloUrl *start_url;

   /* Relative path to a local file? */
   p = (*str == '/') ? dStrdup(str) :
                       dStrconcat(dGetcwd(), "/", str, NULL);

   if (access(p, F_OK) == 0) {
      /* absolute path may have non-URL characters */
      url_str = a_Misc_escape_chars(p, "% ");
#ifdef HAVE_DRIVE_LETTERS
      start_url = a_Url_new(url_str, "file:/");
#else
      start_url = a_Url_new(url_str + 1, "file:/");
#endif
   } else {
      /* Not a file, filter URL string */
      url_str = a_Url_string_strip_delimiters(str);
      start_url = a_Url_new(url_str, NULL);
   }
   dFree(p);
   dFree(url_str);

   if (local)
      a_Url_set_flags(start_url, URL_FLAGS(start_url) | URL_SpamSafe);

   return start_url;
}

/*
 * MAIN
 */
int main(int argc, char **argv)
{
   uint_t opt_id;
   uint_t options_got = 0;
   uint32_t xid = 0;
   int idx = 1;
   int xpos = PREFS_GEOMETRY_DEFAULT_XPOS, ypos = PREFS_GEOMETRY_DEFAULT_YPOS,
       width = PREFS_GEOMETRY_DEFAULT_WIDTH,
       height = PREFS_GEOMETRY_DEFAULT_HEIGHT;
   char **opt_argv;
   FILE *fp;

   srand((uint_t)(time(0) ^ getpid()));

#ifdef SIGPIPE
   // Some OSes exit dillo without this (not GNU/Linux).
   signal(SIGPIPE, SIG_IGN);
#endif /* SIGPIPE */

#ifdef _WIN32
   // Make sure we have unicows.dll for Unicode support on Windows 95/98/Me.
   // Note: This needs to be called before any other modules.
   a_Unicows_check();
#endif /* _WIN32 */

   /* Handle command line options */
   opt_argv = dNew0(char*, numOptions(Options) + 1);
   while ((opt_id = getCmdOption(Options, argc, argv, opt_argv, &idx))) {
      options_got |= opt_id;
      switch (opt_id) {
      case DILLO_CLI_FULLWINDOW:
      case DILLO_CLI_LOCAL:
         break;
      case DILLO_CLI_XID:
      {
         char *end;
         xid = strtol(opt_argv[0], &end, 0);
         if (*end) {
            fprintf(stderr, "XID argument \"%s\" not valid.\n",opt_argv[0]);
            return 2;
         }
         break;
      }
      case DILLO_CLI_GEOMETRY:
         if (!a_Misc_parse_geometry(opt_argv[0],&xpos,&ypos,&width,&height)){
            fprintf(stderr, "geometry argument \"%s\" not valid. Must be of "
                            "the form WxH[{+-}X{+-}Y].\n", opt_argv[0]);
            return 2;
         }
         break;
      case DILLO_CLI_VERSION:
         puts("Dillo version " VERSION);
         return 0;
      case DILLO_CLI_HELP:
         printHelp(argv[0], Options);
         return 0;
      default:
         printHelp(argv[0], Options);
         return 2;
      }
   }
   dFree(opt_argv);

   // set the default values for the preferences
   a_Prefs_init();

   // create ~/.dillo if not present
   Paths::init();

   // initialize default key bindings
   Keys::init();

   // parse dillorc
   if ((fp = Paths::getPrefsFP(PATHS_RC_PREFS))) {
      PrefsParser::parse(fp);
   }
   // parse keysrc
   if ((fp = Paths::getPrefsFP(PATHS_RC_KEYS))) {
      Keys::parse(fp);
   }
   dLib_show_messages(prefs.show_msg);

   // initialize internal modules
   a_Sock_init();
   a_Dns_init();
   a_Web_init();
   a_Http_init();
   a_Mime_init();
   a_Capi_init();
   a_Dicache_init();
   a_Bw_init();
   a_Cookies_init();
   a_Bookmarks_init();
   a_Auth_init();
   a_Download_init();

   /* command line options override preferences */
   if (options_got & DILLO_CLI_FULLWINDOW)
      prefs.fullwindow_start = TRUE;
   if (options_got & DILLO_CLI_GEOMETRY) {
       prefs.width = width;
       prefs.height = height;
       prefs.xpos = xpos;
       prefs.ypos = ypos;
   }

   // Sets WM_CLASS hint on X11
   Fl_Window::default_xclass("dillo");

   if (!prefs.show_tooltip) {
      // turn off UI tooltips
      Fl::option(Fl::OPTION_SHOW_TOOLTIPS, false);
   }

   // Disable '@' and '&' interpretation in FL_FREE_LABELTYPE.
   Fl::set_labeltype(FL_FREE_LABELTYPE, custLabelDraw, custLabelMeasure);

   // Custom box type for CustTabButton (see uicmd.cc).
   Fl::set_boxtype(FL_DIAMOND_UP_BOX, custTabButtonUpBoxDraw, 0, 0, 1, 0);
   Fl::set_boxtype(FL_DIAMOND_DOWN_BOX, custTabButtonDownBoxDraw, 0, 0, 1, 0);

   checkPreferredFonts();
   Fl::get_system_colors();

   // Create a new UI/bw pair
   BrowserWindow *bw = a_UIcmd_browser_window_new(0, 0, xid, NULL);

   /* We need this so that fl_text_extents() in dw/fltkplatform.cc can
    * work when FLTK is configured without XFT and Dillo is opening
    * immediately-available URLs from the cmdline (e.g. about:splash).
    */
   ((Fl_Widget *)bw->ui)->window()->make_current();

   /* Proxy authentication */
   if (prefs.http_proxyuser && !a_Http_proxy_auth()) {
      const char *passwd = a_UIcmd_get_passwd(prefs.http_proxyuser);
      if (passwd) {
         a_Http_set_proxy_passwd(passwd);
      } else {
         MSG_WARN("Not using proxy authentication.\n");
      }
   }

   /* Open URLs/files */
   const bool local = options_got & DILLO_CLI_LOCAL;

   if (idx == argc) {
      /* No URLs/files on cmdline. Send startup screen */
      if (strcmp(URL_STR(prefs.start_page), "about:blank") == 0)
         a_UIcmd_open_url(bw, NULL);
      else
         a_UIcmd_open_url(bw, prefs.start_page);
   } else {
      for (int i = idx; i < argc; i++) {
         DilloUrl *start_url = makeStartUrl(argv[i], local);

         if (i > idx) {
            if (prefs.middle_click_opens_new_tab) {
               /* user must prefer tabs */
               const int focus = 1;
               a_UIcmd_open_url_nt(bw, start_url, focus);
            } else {
               a_UIcmd_open_url_nw(bw, start_url);
            }
         } else {
            a_UIcmd_open_url(bw, start_url);
         }
         a_Url_free(start_url);
      }
   }

   Fl::run();

   /*
    * Memory deallocating routines
    * (This can be left to the OS, but we'll do it, with a view to test
    *  and fix our memory management)
    */
   a_Download_freeall();
   a_Cookies_freeall();
   a_Bookmarks_freeall();
   a_Cache_freeall();
   a_Dicache_freeall();
   a_Http_freeall();
   a_Dns_freeall();
   a_History_freeall();
   a_Prefs_freeall();
   Keys::free();
   Paths::free();
   a_Sock_freeall();
   /* TODO: auth, css */

   MSG("Dillo: normal exit!\n");
   return 0;
}
